src/ralph/virtual/management/commands/openstack_sync.py

```diff
@@ -0,0 +1,532 @@
+    CloudHost,
+    CloudProject,
+    CloudProvider
+)
+
+setup()
+logger = logging.getLogger(__name__)
+
+try:
+    from keystoneclient.v2_0 import client as ks
+    keystone_client_exists = True
+except ImportError:
+    keystone_client_exists = False
+
+try:
+    from novaclient import client as novac
+    nova_client_exists = True
+except ImportError:
+    nova_client_exists = False
+
+
+class Command(BaseCommand):
+    def __init__(self):
+        super().__init__()
+        self.DATETIME_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
+        self.summary = defaultdict(int)
+        self.openstack_projects = {}
+        self.openstack_flavors = {}
+
+    @staticmethod
+    def _get_novaclient_connection(site):
+        nt = novac.Client(
+            site['version'],
+            site['username'],
+            site['password'],
+            site['tenant_name'],
+            site['auth_url']
+        )
+        return nt
+
+    @staticmethod
+    def _get_servers_list(nt, site):
+        """
+        Returns list of servers for a project.
+        :parm site: novaclient connection
+        """
+        servers = []
+        marker = None
+
+        while True:
+            try:
+                servers_part = nt.servers.list(
+                    search_opts={'all_tenants': True},
+                    limit=1000,
+                    marker=marker,
+                )
+                marker = servers_part[-1].id
+                servers.extend(servers_part)
+            except IndexError:
+                break
+
+        servers = list(map(lambda server: server.__dict__, servers))
+        if len(servers) == 0:
+            raise NotImplementedError('Got an empty list of instances from '
+                                      '{}'.format(site['auth_url']))
+        return servers
+
+    @staticmethod
+    def _get_flavors_list(nt, site):
+        """
+        Return list of flavours
+        :parm nt: novaclient connection
+        :return: dict
+        """
+        flavors = list(map(lambda flavor: flavor.__dict__, nt.flavors.list()))
+
+        if len(flavors) == 0:
+            raise NotImplementedError('Got an empty list of flavors from '
+                                      '{}'.format(site['auth_url']))
+        return flavors
+
+    @staticmethod
+    def _get_projects_map(site):
+        """
+        Returns a hash map tenant_id->tenant_name
+        :rtype: dict
+        """
+        keystone_client = ks.Client(
+            username=site['username'],
+            password=site['password'],
+            tenant_name=site['tenant_name'],
+            auth_url=site['auth_url'],
+        )
+
+        projects = {}
+        for project in keystone_client.tenants.list():
+            projects[project.id] = {
+                'name': project.name,
+                'tag':  site['tag'],
+                'servers': {},
+            }
+
+        if len(projects) == 0:
+            raise NotImplementedError('Got an empty list of projects from '
```

Comment:
there should be exception here IMO - we could log this, but it's possible (especially for some test instances)